index/scorch/snapshot_index.go

```diff
@@ -496,30 +496,55 @@ func docInternalToNumber(in index.IndexInternalID) (uint64, error) {
 
 func (i *IndexSnapshot) DocumentVisitFieldTerms(id index.IndexInternalID,
 	fields []string, visitor index.DocumentFieldTermVisitor) error {
+	_, err := i.documentVisitFieldTerms(id, fields, visitor, nil)
+	return err
+}
+
+func (i *IndexSnapshot) documentVisitFieldTerms(id index.IndexInternalID,
+	fields []string, visitor index.DocumentFieldTermVisitor, dvs segment.DocVisitState) (
+	segment.DocVisitState, error) {
 
 	docNum, err := docInternalToNumber(id)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	segmentIndex, localDocNum := i.segmentIndexAndLocalDocNumFromGlobal(docNum)
 	if segmentIndex >= len(i.segment) {
-		return nil
+		return nil, nil
 	}
 
 	ss := i.segment[segmentIndex]
 
 	if zaps, ok := ss.segment.(segment.DocumentFieldTermVisitable); ok {
+		if ss.dvState == nil {
+			ss.dvState = &fieldDocValueState{}
+		}
+
+		// if all fields are dv persisted
+		if ss.dvState.dvFieldsAllPersisted {
```

Comment:
still wondering about this dvFieldsAllPersisted bool?  that is, why there should be a special case for all fields getting persisted (as opposed to, say, some or none of the fields getting persisted)... that is all cases ought to perhaps just "fall out" for free?

Reply:
Added this bool to save further calls to VisitableDocValueFields and extractDvPendingFields on every invocation.